oshmem/mca/spml/ucx/spml_ucx.c

```diff
@@ -126,61 +127,120 @@ int mca_spml_ucx_del_procs(ompi_proc_t** procs, size_t nprocs)
 
     ret = opal_common_ucx_del_procs_nofence(del_procs, nprocs, oshmem_my_proc_id(),
                                             mca_spml_ucx.num_disconnect,
-                                            mca_spml_ucx_ctx_default.ucp_worker);
+                                            mca_spml_ucx_ctx_default.ucp_worker[0]);
     /* No need to barrier here - barrier is called in _shmem_finalize */
     free(del_procs);
-    free(mca_spml_ucx.remote_addrs_tbl);
+    if (mca_spml_ucx.remote_addrs_tbl) {
+        for (w = 0; w < ucp_workers; w++) {
+            if (mca_spml_ucx.remote_addrs_tbl[w]) {
+                for (n = 0; n < nprocs; n++) {
+                    if (mca_spml_ucx.remote_addrs_tbl[w][n]) {
+                        free(mca_spml_ucx.remote_addrs_tbl[w][n]);
+                    }
+                }
+                free(mca_spml_ucx.remote_addrs_tbl[w]);
+            }
+        }
+        free(mca_spml_ucx.remote_addrs_tbl);
+    }
+
     free(mca_spml_ucx_ctx_default.ucp_peers);
 
     mca_spml_ucx_ctx_default.ucp_peers = NULL;
 
     return ret;
 }
 
+static void dump_address(char *s, int pe, char *addr, size_t len)
+{
+#ifdef SPML_UCX_DEBUG
+    int my_rank = oshmem_my_proc_id();
+    unsigned  i;
+
+    printf("%s e=%d dest_pe=%d addr=%p len=%d\n", s, my_rank, pe, addr, len);
+    for (i = 0; i < len; i++) {
+        printf("%02X ", (unsigned)0xFF&addr[i]);
+    }
+    printf("\n");
+#endif
+}
+
 /* TODO: move func into common place, use it with rkey exchng too */
 static int oshmem_shmem_xchng(
-        void *local_data, int local_size, int nprocs,
-        void **rdata_p, int **roffsets_p, int **rsizes_p)
+        void **local_data, unsigned int *local_size, int nprocs, int ucp_workers,
+        void **rdata_p, unsigned int **roffsets_p, unsigned int **rsizes_p)
 {
-    int *rcv_sizes   = NULL;
-    int *rcv_offsets = NULL; 
-    void *rcv_buf    = NULL;
+    unsigned int *rcv_sizes   = NULL;
+    int *_rcv_sizes = NULL;
+    unsigned int *rcv_offsets = NULL;
+    int *_rcv_offsets = NULL; 
+    void *rcv_buf       = NULL;
     int rc;
-    int i;
+    int i,j,k;
 
-    /* do llgatherv */
-    rcv_offsets = malloc(nprocs * sizeof(*rcv_offsets));
+    /* do allgatherv */
+    rcv_offsets = calloc(ucp_workers * nprocs, sizeof(*rcv_offsets));
     if (NULL == rcv_offsets) {
         goto err;
     }
 
     /* todo: move into separate function. do allgatherv */
-    rcv_sizes = malloc(nprocs * sizeof(*rcv_sizes));
+    rcv_sizes = calloc(ucp_workers * nprocs, sizeof(*rcv_sizes));
     if (NULL == rcv_sizes) {
         goto err;
     }
-    
-    rc = oshmem_shmem_allgather(&local_size, rcv_sizes, sizeof(int));
+   
+    rc = oshmem_shmem_allgather(local_size, rcv_sizes, ucp_workers * sizeof(*rcv_sizes));
```

Comment:
Something to consider: this scheme might impact job launch time especially for single thread case. Basically you do `ucp_workers` allgathers. Right?